File: code_puppy/compaction/tool_arg_truncation.py
"""Pre-summarization optimization: truncate large tool call args and returns.

This is a cheap pass that often reclaims significant tokens without needing an
LLM call for full summarization. Targets write_file/edit_file tool calls where
the 'content' arg can be huge, and tool returns (read_file, grep, list_files)
where the return content can be very large.
"""

import dataclasses
import logging
from typing import Any, Sequence

from pydantic_ai.messages import (
    ModelMessage,
    ModelRequest,
    ModelResponse,
    ToolCallPart,
    ToolReturnPart,
)

logger = logging.getLogger(__name__)

# Tool names whose args are commonly large (content, old_string, new_string)
_TARGET_TOOLS: frozenset[str] = frozenset(
    {
        "write_file",
        "edit_file",
        "replace_in_file",
        "create_file",
        "apply_patch",
    }
)

# Keys within a tool call's args that should be truncated if oversized
_TARGET_KEYS: frozenset[str] = frozenset(
    {
        "content",
        "new_content",
        "old_string",
        "new_string",
        "patch",
        "text",
    }
)

DEFAULT_MAX_ARG_LENGTH = 500  # chars
DEFAULT_TRUNCATION_TEXT = " ...(argument truncated during compaction)"

# Tool return truncation defaults
DEFAULT_MAX_RETURN_LENGTH = 5000  # chars
DEFAULT_RETURN_HEAD_CHARS = 500  # chars to keep from start
DEFAULT_RETURN_TAIL_CHARS = 200  # chars to keep from end
DEFAULT_RETURN_TRUNCATION_TEXT = "\n...(truncated during compaction)...\n"


def truncate_tool_arg(
    value: Any,
    max_length: int = DEFAULT_MAX_ARG_LENGTH,
    truncation_text: str = DEFAULT_TRUNCATION_TEXT,
) -> tuple[Any, bool]:
    """Truncate a single arg value if it's a string and too long.

    Args:
        value: The argument value to potentially truncate
        max_length: Maximum length before truncation
        truncation_text: Text to append after truncation

    Returns:
        (new_value, was_modified)
    """
    if not isinstance(value, str):
        return value, False
    if len(value) <= max_length:
        return value, False
    return value[:max_length] + truncation_text, True


def truncate_tool_call_args(
    tool_name: str,
    args: dict[str, Any],
    *,
    max_length: int = DEFAULT_MAX_ARG_LENGTH,
    truncation_text: str = DEFAULT_TRUNCATION_TEXT,
    target_tools: frozenset[str] = _TARGET_TOOLS,
    target_keys: frozenset[str] = _TARGET_KEYS,
) -> tuple[dict[str, Any], bool]:
    """Truncate target keys in a tool call's args if it's a target tool.

    Args:
        tool_name: Name of the tool being called
        args: The arguments dict for the tool call
        max_length: Maximum length before truncation
        truncation_text: Text to append after truncation
        target_tools: Set of tool names to target for truncation
        target_keys: Set of argument keys to target for truncation

    Returns:
        (new_args_dict, was_modified)
    """
    if tool_name not in target_tools:
        return args, False

    new_args: dict[str, Any] = {}
    any_modified = False
    for key, value in args.items():
        if key in target_keys:
            new_value, modified = truncate_tool_arg(value, max_length, truncation_text)
            new_args[key] = new_value
            any_modified = any_modified or modified
        else:
            new_args[key] = value
    return new_args, any_modified


def _truncate_message_tool_calls(
    msg: ModelMessage,
    max_length: int,
) -> ModelMessage:
    """Return a new message with tool call args truncated, or the original if unchanged.

    Handles pydantic-ai's ModelResponse (which contains ToolCallPart) and
    ModelRequest (which can contain ToolCallPart when being re-sent).

    Args:
        msg: The message to process
        max_length: Maximum characters per target argument

    Returns:
        New message with truncated args, or original if no changes needed
    """
    if not isinstance(msg, (ModelResponse, ModelRequest)):
        return msg

    parts = getattr(msg, "parts", None)
    if not parts:
        return msg

    new_parts = []
    any_modified = False

    for part in parts:
        if isinstance(part, ToolCallPart):
            tool_name = getattr(part, "tool_name", "")
            args = getattr(part, "args", {}) or {}

            new_args, modified = truncate_tool_call_args(
                tool_name, args, max_length=max_length
            )

            if modified:
                new_part = dataclasses.replace(part, args=new_args)
                new_parts.append(new_part)
                any_modified = True
            else:
                new_parts.append(part)
        else:
            new_parts.append(part)

    if not any_modified:
        return msg

    if isinstance(msg, ModelResponse):
        return ModelResponse(parts=new_parts)
    elif isinstance(msg, ModelRequest):
        return ModelRequest(parts=new_parts)

    return msg


def truncate_tool_return_content(
    content: Any,
    *,
    max_length: int = DEFAULT_MAX_RETURN_LENGTH,
    head_chars: int = DEFAULT_RETURN_HEAD_CHARS,
    tail_chars: int = DEFAULT_RETURN_TAIL_CHARS,
    truncation_text: str = DEFAULT_RETURN_TRUNCATION_TEXT,
) -> tuple[Any, bool]:
    """Truncate tool return content if it's a string and too long.

    Preserves the first head_chars and last tail_chars characters with a
    truncation marker in between.

    Args:
        content: The tool return content to potentially truncate
        max_length: Total length threshold before truncation kicks in
        head_chars: Characters to keep from the start
        tail_chars: Characters to keep from the end
        truncation_text: Marker text inserted at the truncation point

    Returns:
        (new_content, was_modified)
    """
    if not isinstance(content, str):
        return content, False
    if len(content) <= max_length:
        return content, False
    total_orig = len(content)
    truncated = content[:head_chars] + truncation_text + content[-tail_chars:]
    summary_header = f"[Truncated: tool return was {total_orig} chars]\n"
    return summary_header + truncated, True


def _truncate_message_tool_returns(
    msg: ModelMessage,
    max_length: int = DEFAULT_MAX_RETURN_LENGTH,
    head_chars: int = DEFAULT_RETURN_HEAD_CHARS,
    tail_chars: int = DEFAULT_RETURN_TAIL_CHARS,
) -> ModelMessage:
    """Return a new message with tool return content truncated, or the original.

    Processes ModelRequest parts containing ToolReturnPart instances.

    Args:
        msg: The message to process
        max_length: Maximum characters before truncation kicks in
        head_chars: Characters to keep from the start of the return
        tail_chars: Characters to keep from the end of the return

    Returns:
        New message with truncated returns, or original if no changes needed
    """
    if not isinstance(msg, ModelRequest):
        return msg

    parts = getattr(msg, "parts", None)
    if not parts:
        return msg

    new_parts = []
    any_modified = False

    for part in parts:
        if isinstance(part, ToolReturnPart):
            content = part.content
            new_content, modified = truncate_tool_return_content(
                content,
                max_length=max_length,
                head_chars=head_chars,
                tail_chars=tail_chars,
            )
            if modified:
                new_part = dataclasses.replace(part, content=new_content)
                new_parts.append(new_part)
                any_modified = True
            else:
                new_parts.append(part)
        else:
            new_parts.append(part)

    if not any_modified:
        return msg

    return ModelRequest(parts=new_parts)


def pretruncate_messages(
    messages: Sequence[ModelMessage],
    *,
    keep_recent: int = 10,
    max_length: int = DEFAULT_MAX_ARG_LENGTH,
    max_return_length: int = DEFAULT_MAX_RETURN_LENGTH,
    return_head_chars: int = DEFAULT_RETURN_HEAD_CHARS,
    return_tail_chars: int = DEFAULT_RETURN_TAIL_CHARS,
) -> tuple[list[ModelMessage], int]:
    """Pre-truncate tool call args AND tool returns in older messages.

    This is an OPTIONAL cheap pass to run BEFORE full summarization. It tries to
    reclaim tokens without an LLM call.

    Args:
        messages: Full message history (pydantic-ai messages)
        keep_recent: Don't touch the last N messages (they're active context)
        max_length: Max characters allowed per target arg before truncation
        max_return_length: Max characters for tool returns before truncation
        return_head_chars: Characters to keep from start of truncated returns
        return_tail_chars: Characters to keep from end of truncated returns

    Returns:
        (modified_messages, truncation_count) -- modified_messages is a new list;
        the original is not mutated.
    """
    if len(messages) <= keep_recent:
        return list(messages), 0

    if keep_recent > 0:
        older = list(messages[:-keep_recent])
        recent = list(messages[-keep_recent:])
    else:
        older = list(messages)
        recent = []

    truncation_count = 0
    modified_older: list[ModelMessage] = []

    for msg in older:
        # First pass: truncate tool call args
        new_msg = _truncate_message_tool_calls(msg, max_length)
        # Second pass: truncate tool return content
        new_msg = _truncate_message_tool_returns(
            new_msg,
            max_length=max_return_length,
            head_chars=return_head_chars,
            tail_chars=return_tail_chars,
        )
        if new_msg is not msg:
            truncation_count += 1
        modified_older.append(new_msg)

    return modified_older + recent, truncation_count

File: lib/code_puppy_control/compaction/tool_arg_truncation.ex
defmodule CodePuppyControl.Compaction.ToolArgTruncation do
  @moduledoc """
  Pre-summarization optimization: truncate large tool call args and results.

  This is a cheap pass that reclaims significant tokens without an LLM call.
  Targets write_file/edit_file tool calls where the 'content' arg can be huge,
  and tool returns (read_file, grep, list_files) where return content can be very large.

  Port of `code_puppy/compaction/tool_arg_truncation.py`.

  ## Usage

      {messages, count} = ToolArgTruncation.pretruncate_messages(messages, keep_recent: 10)
  """

  # Tool names whose args are commonly large
  @target_tools MapSet.new(~w(write_file edit_file replace_in_file create_file apply_patch))

  # Keys within tool call args to truncate if oversized
  @target_keys MapSet.new(~w(content new_content old_string new_string patch text))

  # Defaults
  @default_max_arg_length 500
  @default_truncation_text " ...(argument truncated during compaction)"
  @default_max_return_length 5000
  @default_return_head_chars 500
  @default_return_tail_chars 200
  @default_return_truncation_text "\n...(truncated during compaction)...\n"

  @doc """
  Truncate a single string value if it exceeds max_length.

  Returns `{new_value, was_modified}`.
  """
  @spec truncate_value(String.t(), pos_integer(), String.t()) :: {String.t(), boolean()}
  def truncate_value(
        value,
        max_length \\ @default_max_arg_length,
        truncation_text \\ @default_truncation_text
      )

  def truncate_value(value, max_length, truncation_text) when is_binary(value) do
    if String.length(value) <= max_length do
      {value, false}
    else
      {String.slice(value, 0, max_length) <> truncation_text, true}
    end
  end

  def truncate_value(value, _max_length, _truncation_text), do: {value, false}

  @doc """
  Truncate target keys in a tool call's args if it's a target tool.

  Returns `{new_args_map, was_modified}`.
  """
  @spec truncate_tool_call_args(String.t(), map() | any(), keyword()) :: {any(), boolean()}
  def truncate_tool_call_args(tool_name, args, opts \\ [])

  def truncate_tool_call_args(tool_name, args, opts) when is_map(args) do
    if MapSet.member?(@target_tools, tool_name) do
      max_length = Keyword.get(opts, :max_length, @default_max_arg_length)
      truncation_text = Keyword.get(opts, :truncation_text, @default_truncation_text)
      target_keys = Keyword.get(opts, :target_keys, @target_keys)

      {new_args, any_modified} =
        Enum.reduce(args, {%{}, false}, fn {key, value}, {acc, modified?} ->
          if MapSet.member?(target_keys, key) do
            {new_value, was_truncated} = truncate_value(value, max_length, truncation_text)
            {Map.put(acc, key, new_value), modified? or was_truncated}
          else
            {Map.put(acc, key, value), modified?}
          end
        end)

      {new_args, any_modified}
    else
      {args, false}
    end
  end

  def truncate_tool_call_args(_tool_name, args, _opts), do: {args, false}

  @doc """
  Truncate tool return content if it's a string and too long.

  Preserves the first head_chars and last tail_chars with a truncation marker.

  Returns `{new_content, was_modified}`.
  """
  @spec truncate_tool_return_content(any(), keyword()) :: {any(), boolean()}
  def truncate_tool_return_content(content, opts \\ [])

  def truncate_tool_return_content(content, opts) when is_binary(content) do
    max_length = Keyword.get(opts, :max_length, @default_max_return_length)
    head_chars = Keyword.get(opts, :head_chars, @default_return_head_chars)
    tail_chars = Keyword.get(opts, :tail_chars, @default_return_tail_chars)
    truncation_text = Keyword.get(opts, :truncation_text, @default_return_truncation_text)

    if String.length(content) <= max_length do
      {content, false}
    else
      total = String.length(content)
      head = String.slice(content, 0, head_chars)
      tail = String.slice(content, -tail_chars, tail_chars)
      truncated = head <> truncation_text <> tail
      {"[Truncated: tool return was #{total} chars]\n" <> truncated, true}
    end
  end

  def truncate_tool_return_content(content, _opts), do: {content, false}

  @doc """
  Pre-truncate tool call args AND tool returns in older messages.

  This is an OPTIONAL cheap pass to run BEFORE full summarization. It tries to
  reclaim tokens without an LLM call.

  ## Options

    * `:keep_recent` — Don't touch the last N messages (default: 10)
    * `:max_length` — Max characters per target arg (default: 500)
    * `:max_return_length` — Max characters for tool returns (default: 5000)
    * `:return_head_chars` — Characters to keep from start of returns (default: 500)
    * `:return_tail_chars` — Characters to keep from end of returns (default: 200)

  Returns `{modified_messages, truncation_count}`.
  """
  @spec pretruncate_messages([map()], keyword()) :: {[map()], non_neg_integer()}
  def pretruncate_messages(messages, opts \\ []) when is_list(messages) do
    keep_recent = Keyword.get(opts, :keep_recent, 10)

    if length(messages) <= keep_recent do
      {messages, 0}
    else
      {older, recent} = Enum.split(messages, length(messages) - keep_recent)

      {modified_older, count} =
        Enum.reduce(older, {[], 0}, fn msg, {acc, trunc_count} ->
          new_msg = truncate_message_parts(msg, opts)

          if new_msg != msg do
            {[new_msg | acc], trunc_count + 1}
          else
            {[msg | acc], trunc_count}
          end
        end)

      {Enum.reverse(modified_older) ++ recent, count}
    end
  end

  @doc """
  Truncate a single message's parts in-place.

  Processes both tool-call args and tool-return content.
  """
  @spec truncate_message_parts(map(), keyword()) :: map()
  def truncate_message_parts(message, opts \\ []) do
    parts = get_field(message, :parts) || []

    {new_parts, any_modified?} =
      Enum.reduce(parts, {[], false}, fn part, {acc, modified?} ->
        {new_part, was_modified} = truncate_part(part, opts)
        {[new_part | acc], modified? or was_modified}
      end)

    if any_modified? do
      put_field(message, :parts, Enum.reverse(new_parts))
    else
      message
    end
  end

  # --- Private ---

  defp truncate_part(part, opts) do
    kind = get_field(part, :part_kind)

    cond do
      kind == "tool-call" ->
        truncate_tool_call_part(part, opts)

      kind == "tool-return" or kind == "tool_return" ->
        truncate_tool_return_part(part, opts)

      true ->
        {part, false}
    end
  end

  defp truncate_tool_call_part(part, opts) do
    tool_name = get_field(part, :tool_name) || ""
    args = get_field(part, :args) || %{}

    {new_args, modified?} = truncate_tool_call_args(tool_name, args, opts)

    if modified? do
      {put_field(part, :args, new_args), true}
    else
      {part, false}
    end
  end

  defp truncate_tool_return_part(part, opts) do
    content = get_field(part, :content)

    {new_content, modified?} = truncate_tool_return_content(content, opts)

    if modified? do
      {put_field(part, :content, new_content), true}
    else
      {part, false}
    end
  end

  # Handle both atom and string keys for cross-module compatibility
  defp get_field(map, key) when is_atom(key) do
    case Map.get(map, key) do
      nil -> Map.get(map, Atom.to_string(key))
      val -> val
    end
  end

  defp put_field(map, key, value) when is_atom(key) do
    if Map.has_key?(map, key) do
      Map.put(map, key, value)
    else
      Map.put(map, Atom.to_string(key), value)
    end
  end
end
